Document APNs and ABM setup and renewal (#9471)
fleetdm/fleet: docs/Using-Fleet/Mobile-device-management.md

MDM features are not ready for production and are currently in development. These features are disabled by default.

MDM features allow you to manage macOS updates and macOS settings on your hosts.

> ## Apple Push Notification Service (APNs)

Reviewer: ```suggestion
To use MDM features you have to connect Fleet to Apple Push Certificates Portal. See how [here](#apple-push-notification-service-apns).

## Apple Push Notification Service (APNs)
```

Hey, can we keep the "To use MDM features..." one line summary? This way the top of the page tells the user "you have to do this to use these features."

Then we can link them to the APNs section.

Also, I think we should move APNs and ABM sections to the bottom of the page. Even though they're the first step in setting up MDM features, the user will only follow these instructions once (and 1 year later when they renew). The user is likely to come back to this page to read about the OS updates (macOS updates) and Profiles (macOS settings) features more frequently. What do you think?

Author: :+1: sounds good, I'll update the PR.

> To use MDM features you have to connect Fleet to Apple Push Certificates Portal. This can be done via the Fleet UI or the `fleetctl` command-line interface. The result is the same regardless of the method used, you end up with 4 files: an APNs certificate and private key files along with the Simple Certificate Enrollment Protocol (SCEP) certificate and private key files. Make sure you store them securely as you will need them to configure your Fleet instances afterwards (and take note of the email and organization used to generate them, as you will need to use the same ones during renewal). For _renewal_ of an existing but expired (or soon to expire) APNs certificate, see the [APNs Renewal section](#apns-renewal) below.

Reviewer: ```suggestion
To connect Fleet to Apple, get these four files using the Fleet UI or the `fleetctl` command-line interface: An APNs certificate, APNs private, Simple Certificate Enrollment Protocol (SCEP) certificate, and SCEP private key. 

To do this, choose the "Fleet UI" or "fleetctl" method and follow the steps below.
```

I think we can shorten some of this for clarity. Please let me know if I'm cutting anything important.

I moved removed the "store the certs/keys securely" sentence and moved "take note of the email" sentence below.

> Via the Fleet UI:

Reviewer: ```suggestion
Fleet UI:
```

1. Head to the **Settings > Integrations > Mobile device management (MDM)** page. Users with the admin role can access the settings pages.
2. Follow the instructions under **Apple Push Certificates Portal**.

> Via the `fleetctl` CLI:

Reviewer: ```suggestion
`fleetctl` CLI:
```

1. Run `fleetctl generate mdm-apple --email <email> --org <org>`.

> 2. Follow the on-screen instructions.

Reviewer: ```suggestion
2. Follow the on-screen instructions.

> Take note of the Apple ID you use to sign into Apple Push Certificates Portal. You'll need to use the same Apple ID when renewing your APNs certificate. Apple requires that APNs certificates are renewed once every year. To renew, see the [APNs Renewal section](#apns-renewal) .
```

@mna do you know if the `--email` and `--org` matter? My understanding is that the Apple ID used to create the APNs cert matters.

Author: I think you removed the mention to store the certs/keys securely for use with Fleet's config later on:

> Make sure you store them securely as you will need them to configure your Fleet instances afterwards

Was that on purpose?

Author: > do you know if the --email and --org matter?

Yeah I'm not sure about the --org, the documentation elsewhere only mentions the Apple ID, so that may be the only requirement.

Reviewer: > Yeah I'm not sure about the --org, the documentation elsewhere only mentions the Apple ID, so that may be the only requirement.

Got it. @roperzh do you know? Can we test this? 

For now, let's merge this as (Apple ID is the only requirement). If we're wrong we can come back and change.

Reviewer: > I think you removed the mention to store the certs/keys securely for use with Fleet's config later on

@mna this was intentional because I think it's implied. As a user following the instructions, I generate the certs/keys and then the next step (or the step after) is configuring them.

Please feel free to add it back if you disagree (you think it's not implied).

Author: > As a user following the instructions, I generate the certs/keys and then the next step (or the step after) is configuring them.
> 
> Please feel free to add it back if you disagree (you think it's not implied).

That's probably fine, that was just as an extra precaution, but I like how terse your proposals are.

Reviewer 2: @noahtalerman @mna the org it's not required to generate the APNs key pair, but it is required by `fleetctl`, this was a requirement in the issue.

This was also briefly discussed here: https://github.com/fleetdm/fleet/pull/8812#discussion_r1035394426

### APNs Renewal

> The APNs certificate is typically valid for a year. You can see the certificate's renewal date and other important APNs information via the Fleet UI or the `fleetctl` command-line interface:

Reviewer: ```suggestion
Apple requires that APNs certificates are renewed once every year. You can see the certificate's renewal date and other important APNs information using the Fleet UI or the `fleetctl` command-line interface:
```

> Via the Fleet UI:

Reviewer: ```suggestion
Fleet UI:
```

1. Head to the **Settings > Integrations > Mobile device management (MDM)** page. Users with the admin role can access the settings pages.
2. Look at the **Apple Push Certificates Portal** section.

> Via the `fleetctl` CLI:

Reviewer: ```suggestion
`fleetctl` CLI:
```

1. Run `fleetctl get mdm-apple`.
2. Look at the on-screen information.

> If the certificate is expired or about to expire, you must renew it by using the same command that can be used to generate it the first time, `fleetctl generate mdm-apple --email <email> --org <org>`. The Fleet UI cannot be used for renewal at the moment. Note that you must make sure that you use the same email and organization as when the certificate was generated. One **important difference** when renewing a certificate is that in the [Apple portal to get the new certificate](https://identity.apple.com), you must click on the _Renew_ button so that the same APNs topic is reused.

Reviewer: ```suggestion
How to renew the certificate if it's expired or about to expire: 

1. Run the `fleetctl generate mdm-apple --email <email> --org <org>` command. 

2. Sign in to [Apple Push Certificates Portal](https://identity.apple.com) using the same Apple ID you used to get your original certificate. If you don't use the same Apple ID, you will have to turn MDM off and back on for all macOS hosts.

3. In the **Settings > Integrations > Mobile device management (MDM)** page, under Apple Push Certificates portal, find the serial number of your current certificate. In Apple Push Certificates Portal, click  **Renew** next to the certificate that has the matching serial number. If you don't renew and get a new certificate, you will have to turn MDM off and back on for all macOS hosts.
```

I broke this out into several steps and added information about finding the serial number in the Fleet UI.

I have the same question about `--org` and `--email`. Does using the same matter?

Author: I'm sure for --email, not for --org, your text looks good (and clearer) to me.

Reviewer: > I'm sure for --email, not for --org

@mna hmm, I'm confused. 

Let's say I use `noah@example.com` to generate my original CSR. I log in to Apple Push Certs Portal with this email. Later, when I renew, I set `--email` to `bob@example.com` I still login to Apple with `noah@example.com`.

Will renewing work? (I don't have to re-enroll hosts).

Or, during renewal, do I have to use `noah@example.com` to generate the CSR? To prevent having to re-enroll hosts.

Reviewer: The below is from the following Slack thread (internal): https://fleetdm.slack.com/archives/C03C41L5YEL/p1674586568847999?thread_ts=1674586435.140469&cid=C03C41L5YEL

> Let's say I use noah@example.com to generate my original CSR. I log in to Apple Push Certs Portal with this email. Later, when I renew, I set --email to bob@example.com I still login to Apple with noah@example.com.

Martin

You will probably see the certificate to renew, but my understanding is that after renewal, you may not find the new cert there (it will be in the bob account, or it may even fail to renew, I don't know). According to https://docs.jamf.com/jamf-now/documentation/Determining_the_Apple_ID_Used_for_APNs_Setup.html , the certificate can be found under the account of the certificate owner AFAICT.

All in all, I think it's better to recommend using the same? Though surely there's a process to follow if the old ID is not at the organization anymore.

Noah

> after renewal, you may not find the new cert there (it will be in the bob account, or it may even fail to renew, I don’t know).

Ah, got it. Yeah I’m not sure either. Can we test this?

Noah

For now, adding back the recommendation to use the same email makes sense to me. This way we can get the docs merged

Martin

> Can we test this?

When we QA this we should make sure to cover that. I left a comment to @xpkoala on the ticket, I'll add that part to it (renew with a different email).

## Apple Business Manager (ABM)

_Available in Fleet Premium_

> Apple Business Manager (ABM) supports automatic enrollment and management of devices via Device Enrollment Program (DEP) enrollment. In order to configure Fleet instances with ABM enabled, you need to generate an ABM certificate and private key, create a new MDM server on [Apple's Business Manager website](https://business.apple.com), associate it with the generated public certificate, and download the encrypted ABM server token.

Reviewer: ```suggestion
Connect Fleet to your ABM account to automatically enroll macOS hosts to Fleet when they’re first unboxed.

To connect Fleet to ABM, get these four files using the Fleet UI or the `fleetctl` command-line interface: An ABM certificate, private key and server token.

To do this, choose the "Fleet UI" or "fleetctl" method and follow the steps below.
```

Same proposed change as here:

> At the end of this process, you end up with 3 files: the ABM certificate, the private key and the encrypted server token. As for the APNs setup described above, this can be done via the Fleet website or the `fleetctl` command-line interface.

Reviewer: ```suggestion
```

> Via the Fleet website:

Reviewer: ```suggestion
Fleet UI:
```

1. In the Fleet UI, head to the **Settings > Integrations > Mobile device management (MDM)** page. Users with the admin role can access the settings pages.
2. Follow the instructions under **Apple Business Manager**.

> Via the `fleetctl` CLI:

Reviewer: ```suggestion
`fleetctl` CLI:
```

1. Run `fleetctl generate mdm-apple-bm`.
2. Follow the on-screen instructions.

### ABM Renewal

The Apple Business Manager server token expires after a year or whenever the account that downloaded the token has their password changed. To renew the token, follow the [instructions documented in this FAQ](https://fleetdm.com/docs/using-fleet/faq#how-can-i-renew-my-apple-business-manager-server-token).

> ## Configuring Fleet instances
> 
> All MDM features need some configuration to be provided to the Fleet instances. All Fleet instances should be configured with the same MDM settings. See https://fleetdm.com/docs/deploying/configuration#mobile-device-management-mdm for all MDM-related configuration options.
> 

Reviewer: ```suggestion
```

I think we can remove this section because it's covered by the instructions in the UI and fleetctl. What do you think?
